src/casing.rs

```diff
@@ -0,0 +1,61 @@
+use onig::Regex;
+
+/// Normalize the casing of a string to be `snake_case`.
+///
+/// This function produces strings that are transformed consistently from a
+/// variety of different input casing. The rule-set for word boundaries are
+/// derived from the withoutboats/heck crate. Underscores are considered word
+/// boundaries in addition to the standard pattern e.g. `\b`. `camelCasing` is
+/// detected by a lowercase followed by an uppercase. Numbers can take on either
+/// case depending on the preceeding symbol.
+///
+/// See: https://github.com/withoutboats/heck/blob/master/src/lib.rs#L7-L17
+pub fn to_snake_case(input: &str) -> String {
+    lazy_static! {
+        static ref EXTRA_SYMBOL: Regex = Regex::new(r"[^\w]|_").unwrap();
+        // This regex matches camelCase in reverse, since the lookbehind
+        // operation only accepts patterns of fixed length. Reversing let's us
+        // determine whether several digits will be uppercase or lowercase.
+        static ref REV_WORD_BOUNDARY: Regex = Regex::new(
+            r"(?x)
+            \b                              # standard word boundary
+            |(?<=[a-z][A-Z])(?=\d*[A-Z])    # break on runs of uppercase e.g. A7Aa -> A7|Aa
+            |(?<=[a-z][A-Z])(?=\d*[a-z])    # break in runs of lowercase e.g a7Aa -> a7|Aa
+            |(?<=[A-Z])(?=\d*[a-z])         # ends with an uppercase e.g. a7A -> a7|A
```

Comment:
I'm a regex fan, but ... oh boy :D 
Did you come up with the regex here or is that copied from somewhere?

Reply:
I [came up with the regex](https://github.com/acmiyaguchi/test-casing/blob/8ca3d68db512fd3a17868c0b08cc84909ebebbc7/src/main.py#L1-L34) in a separate project for testing all of the different casing functions/libraries that we've been using. Setting up the tests was the key result, I have an [auto-generated reporting](https://github.com/acmiyaguchi/test-casing/blob/master/RESULTS.md) for different implementations of snake_casing, including this regex algorithm that's implemented in python.

Here are some other resources that I ended up using:

* [[RexEgg] - Regex Boundaries and Delimiters—Standard and Advanced](https://www.rexegg.com/regex-boundaries.html) - This introduced the concept of word boundaries like `\b` and how to implement it from scratch using lookaheads and lookbehinds.
* [[StackOverflow] - RegEx to split camelCase or TitleCase (advanced)](https://stackoverflow.com/a/7599674) - This validated my approach as I was developing the regex in an online REPL. It does not include digits, which can be either uppercase or lowercase depending on the previous letter.
* [[StackOverflow] - What's the technical reason for “lookbehind assertion MUST be fixed length” in regex?](https://stackoverflow.com/a/40078049) - So how can you tell whether a digit is lowercase or uppercase (e.g. `f("a7aA") == f("A7AA")`)? My first attempt was something similar to `(?<=[a-z]\d*)(?=[A-Z][a-z])` for a "lowercase" digit, but this is not supported by most engines because dynamic look-behinds can perform very badly. I found a neat idea for "inverted" positive lookaheads, which can be used to figure out if a digit is lowercase or uppercase in the 2nd and 3rd lines of the regex.

These three resources, along with the definition of a word boundary and the reference test cases in heck, were useful with developing the regex.